Add IterableOps#tapEach method

joshlemer opened the pull request:
This closes scala/bug#11098

This adds `tapEach` method to `IterableOps` which allows you to write:

```scala
View(1,2,3)
  .tapEach(println)
  .tapEach(println)
  .to(Seq)
// prints: 
1
1
2
2
3
3
```

Or 
```scala
List(1,2,3)
  .tapEach(println)
  .tapEach(println)
// prints: 
1
2
3
1
2
3
```

NthPortal commented on src/library/scala/collection/Iterable.scala:
```diff
@@ -807,6 +807,16 @@ trait IterableOps[+A, +CC[_], +C] extends Any with IterableOnce[A] with Iterable
+    * @tparam U the return type of f
+    * @return The same logical collection as this
+    */
+  def tapEach[U](f: A => U): C = fromSpecific(new View.TapEach(this, f))
```
you could potentially optimise this for strict collections by doing

```scala
  def tapEach[U](f: A => U): C = knownSize match {
    case 0          => this
    case n if n < 0 => fromSpecific(new View.TapEach(this, f))
    case n if n > 0 =>
      foreach(f)
      this
```

NthPortal replied:
it's not perfect (e.g. it misses `List`), but it might be better than nothing?

joshlemer replied:
I think that this size optimization could result in surprising behaviour if the underlying collection is mutated but lazy. In that case, when the empty collection has elements inserted, and then the lazy collection forces, `f` will not be called. I do not know if there's actually any collections like this however...

NthPortal commented on src/library/scala/collection/Iterable.scala:
```diff
@@ -807,6 +807,16 @@ trait IterableOps[+A, +CC[_], +C] extends Any with IterableOnce[A] with Iterable
+    * @tparam U the return type of f
+    * @return The same logical collection as this
+    */
+  def tapEach[U](f: A => U): C = fromSpecific(new View.TapEach(this, f))
```
I'm also not sure if we really need a `View.TapEach`, or can just do `map(a => { f(a); a })` instead of `fromSpecific(new View.TapEach(this, f))`

joshlemer replied:
You're right, I'll change this

joshlemer commented:
@NthPortal ready for re-review

joshlemer commented:
@javax-swing Added some tests [here](https://github.com/scala/scala/pull/7124/files#diff-c8e80d186a09bf13da87d13cb4913e54R376) for LazyList#tapEach

NthPortal commented:
The fact that you need a specialization for `LazyList` is why I think it would be better to implement this as a call to `map`, rather than as `fromSpecific(View.Map(...))`; once you have `map` doing the right thing, `tapEach` will be correct for free

NthPortal commented:
I'm wondering if it would be a good idea to have a `StrictTapEach` trait to mix into strict collections; e.g.

```scala
trait StrictTapEach[A] {
  self: IterableOps[A, AnyConstr, _] =>
  override def tapEach[U](f: A => U): this.type = {
    foreach(f)
    this
  }
}
```

(if `this.type` isn't valid for for some reason, you could add a `C` type parameter to the self type)

joshlemer commented:
>I'm wondering if it would be a good idea to have a StrictTapEach trait to mix into strict collections; 
@NthPortal  
Actually I just put the implementation in `StrictOptimizedIterableOps` which all the strict iterables extend

NthPortal commented on src/library/scala/collection/StrictOptimizedIterableOps.scala:
```diff
@@ -237,4 +237,10 @@ trait StrictOptimizedIterableOps[+A, +CC[_], +C]
+  // Optimization avoids creation of second collection
+  override def tapEach[U](f: A => U): C  = {
+    foreach(f)
+    coll
```
can't return `this`?

joshlemer replied:
Unfortunately, `StrictOptimizedSeqOps[..., C]` does not conform to `C`. I could play around with introducing a self-type to it, `self: C => ` and see if there are any issues.

joshlemer replied:
Ok, It seems that this solution does solve it, as long as you also enforce the `self: C =>` type on all extending `StrictOptimized{Seq,Set,Map,SortedMap, SortedSet}Ops`. I'd like to get @julienrf 's opinion though if he forsees any issues with that constraint, maybe it's intended to be extendable by non-`C` things too

szeiger replied:
The exceptions used to be `StringOps` and `ArrayOps`. All regular collection types extend their `C` parameter.  Can we make any further simplifications if we always enforce this? I don't think it's worth making this change specifically for `tapEach`. There are plenty of other methods that return `C`. In most cases you abstract over consuming a collection where this is not an issue. If you also need to produce a more specific collection type you need to use the proper abstraction anyway (unless we can change it in a more generic way).

joshlemer replied:
@szeiger I can't think of any other improvements or simplifications that could be made if it had `{ self: C => ` unfortunately.

julienrf commented on src/library/scala/collection/StrictOptimizedIterableOps.scala:
```diff
@@ -237,4 +237,10 @@ trait StrictOptimizedIterableOps[+A, +CC[_], +C]
+  // Optimization avoids creation of second collection
+  override def tapEach[U](f: A => U): self.type  = {
+    foreach(f)
+    this
```
I think it would be simpler to return `coll` here (and keep using `C` as the return type).

NthPortal commented on test/junit/scala/collection/immutable/LazyListTest.scala:
```diff
@@ -1000,4 +1001,32 @@ class LazyListTest {
+      assertEquals(Vector(1, 2), ll.take(2).to(Vector))
+      assertEquals(ListBuffer(1, 2), lb)
+      assertEquals(4, ll(3))
+      assertEquals(ListBuffer(1, 2, 3, 4), lb) // this should change to Listbuffer(1,2,4) if LazyList becomes independently lazy
```
you can make the `LazyList` independently lazy by doing `LazyList.tabulate(5)(_ + 1)`

NthPortal commented on test/junit/scala/collection/immutable/LazyListTest.scala:
```diff
@@ -1000,4 +1001,32 @@ class LazyListTest {
       assertEquals(s"mkString 3 $i", goal(3,i), precyc(3,i).mkString)
     }
   }
+  @Test
```
Ideally there should be a test `def tapEach_properlyLazy(): Unit` right after `def map_properlyLazy(): Unit` which asserts laziness; it can be something like the following:

```scala
   @Test
   def tapEach_properlyLazy(): Unit = {
     val op = lazyListOp(_.tapEach(_ + 1)) // or some other invocation of `tapEach` here
     assertRepeatedlyFullyLazy(op)
     assertLazyTailWhenHeadEvaluated(op)
     assertLazyHeadWhenTailEvaluated(op)
     assertLazyHeadWhenNextHeadEvaluated(op)
   }
```

You can also have this method to check that the behaviour is correct, but there are a lot of `assert<Something>` methods you may find helpful for checking laziness (you can look at the other `<op>_properlyLazy()` methods for examples).

joshlemer commented:
@NthPortal your suggestions are implemented now

NthPortal commented on src/library/scala/collection/Iterator.scala:
```diff
@@ -859,6 +859,23 @@ trait Iterator[+A] extends IterableOnce[A] with IterableOnceOps[A, Iterator, Ite
+    * @param f a function to apply to each element in this $coll
+    * @tparam U the return type of f
+    * @return The same logical collection as this
+    */
```
we probably don't need three copies of this scaladoc

NthPortal commented on test/junit/scala/collection/IteratorTest.scala:
```diff
@@ -520,4 +520,27 @@ class IteratorTest {
   }
+
+  @Test def tapEach(): Unit = {
+    {
```
should these use `locally`?

NthPortal reviewed (approved):
Nice!

NthPortal commented:
IMO, this is ready to merge (once the conflict is resolved). @julienrf do you agree?

joshlemer commented:
I think an argument could be made to manually override each strict
collection with an implementation that returns `this.type`, what do you
think?

On Sun., Sep. 16, 2018, 4:04 p.m. Nth, <notifications@github.com> wrote:

> IMO, this is ready to merge (once the conflict is resolved). @julienrf
> <https://github.com/julienrf> do you agree?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/scala/scala/pull/7124#issuecomment-421823673>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AE4rVw15m3hYKhERvQEABGS7alNo6Nn2ks5ubq7ggaJpZM4WIzDe>
> .
>

NthPortal commented:
@joshlemer I'm not sure if I'm behind increasing the maintenance burden by adding a whole bunch of overrides that just call super. If there's a way to do it generically so that you only need at most a few overrides, that sounds good to me, but I'm not sure it's worth it to do it for every strict collection.

szeiger commented on src/library/scala/collection/IterableOnce.scala:
```diff
@@ -429,6 +429,17 @@ trait IterableOnceOps[+A, +CC[_], +C] extends Any { this: IterableOnce[A] =>
     */
   def span(p: A => Boolean): (C, C)
 
+  /** Applies a (typically) side-effecting function to each element in this collection.
```
Is there a non-typical use? I would omit "typically".

joshlemer replied:
👍

SethTisue commented on src/library/scala/collection/immutable/LazyList.scala:
```diff
@@ -368,6 +368,8 @@ final class LazyList[+A] private(private[this] var lazyState: () => LazyList.Sta
     if (knownIsEmpty) LazyList.empty
     else (mapImpl(f): @inline)
 
+  override def tapEach[U](f: A => U): LazyList[A] = map { a => f(a); a}
```
should this be `mapConserve` instead of `map`?

joshlemer replied:
I've never heard of `mapConserve` and it doesn't seem to be available on `LazyList`, what's that?

SethTisue replied:
it avoids consing up a new list if all the items are unchanged

but ohhhh I see `mapConserve` only exists on `List`. maybe `LazyList` could or should have it, nbd, it's pretty obscure stuff, though. it's used a fair amount within the compiler.

nm I guess

NthPortal replied:
I don't think you can have it on `LazyList` - in order to tell if you can return `this`, you need to evaluate the whole mapped list, which violates the laziness

adriaanm commented:
@szeiger do I read your review correctly as "not convinced yet?"

NthPortal commented:
Is there a reason for this not to be merged?

joshlemer commented:
@nthportal I don't believe so

adriaanm commented:
ping @szeiger

szeiger commented on src/library/scala/collection/Iterator.scala:
```diff
@@ -859,6 +859,15 @@ trait Iterator[+A] extends IterableOnce[A] with IterableOnceOps[A, Iterator, Ite
       }
     }
 
+  override def tapEach[U](f: A => U): Iterator[A] = new AbstractIterator[A] {
```
This should override `knownSize`

joshlemer replied:
Done!

szeiger reviewed (commented):
Other than the missing `knownSize`, this looks good.

NthPortal commented on src/library/scala/collection/Iterable.scala:
```diff
@@ -807,6 +807,8 @@ trait IterableOps[+A, +CC[_], +C] extends Any with IterableOnce[A] with Iterable
     */
   def inits: Iterator[C] = iterateUntilEmpty(_.init)
 
+  override def tapEach[U](f: A => U): C = fromSpecific(new View.Map(this, { a: A => f(a); a }))
```
Could I get others' opinions on whether or not this should be implemented as a call to `map` instead of `fromSpecific(new View.Map)`? I think that the former simplifies/reduces the work for those implementing collections which need to override `map`, but I am curious to hear others' thoughts.

I don't think the decision should hold up the PR - it's an easy change to make in its own PR if needed.

julienrf replied:
I think using `map` is better, indeed.

joshlemer replied:
Unfortunately, calling `map { a => f(a); a }` will only return a `CC[A]`, rather than the more specific `C`.

NthPortal replied:
That is interesting and unfortunate. I'm torn on whether or not it would still be better to use `map` and cast the result. I'm leaning towards not.

SethTisue replied:
I asked @szeiger to look at this and explain it to me.

Josh's point is that we can't typecast our way out of this. `CC` and `C[A]` can be different collection types entirely:

```
scala 2.13.0-M5> (Map(1 -> "foo"): Iterable[(Int, String)]).map(x => x)
res5: Iterable[(Int, String)] = List((1,foo))

scala 2.13.0-M5> (collection.immutable.BitSet(0, 5, 9): Iterable[Int]).map(_ + 1)
res0: Iterable[Int] = Set(1, 6, 10)
```

(in an earlier edit, I'd included a `String` example where I "upcast" it to `Iterable[Char]`, but that's not really the same, because it's an implicit conversion, not an upcast)

SethTisue replied:
@szeiger asks whether using `filter` instead was considered. the problem with `map` is that `map` loses the information that the element type is unchanged, but with `filter` we have that

```diff
-  override def tapEach[U](f: A => U): C = fromSpecific(new View.Map(this, { a: A => f(a); a }))
+  override def tapEach[U](f: A => U): C = filter{a: A => f(a); true}
```

neither Stefan and I can think of a downside. there's no significant upside, either, that we can see, it just seems a touch cleaner. thoughts? (the `filter` version compiles, but I ran out of time to try running the tests on it)

joshlemer replied:
@SethTisue Wow!  Very sneaky! Some downsides (which may have workarounds):

* filter can (as far as the compiler is concerned) change the length of the `Iterable`. So the consequences of this are that, perhaps non-strict Seq's like SeqView and LazyList will have to then go through each element in order to fulfill a call to `apply(100)`, whereas before the `filter` call, they may have been able to independently evaluate/access element `100` immediately. 
* The above point means that `IndexedSeqView#filter` cannot return an `IndexedSeqView`

NthPortal replied:
To add on one more point, `filter` also destroys independent laziness of elements (currently only a thing for `LazyList`, but could potentially be a thing for some lazy `IndexedSeq` in the future).

SethTisue replied:
@joshlemer okay, sounds like it’s better as-is then

@NthPortal I don’t follow?

NthPortal replied:
@SethTisue 

```scala
scala> def list = LazyList.tabulate(10)(i => i)
list: scala.collection.immutable.LazyList[Int]

scala> def f(i: Int) = println(i)
f: (i: Int)Unit

scala> list.map(i => {f(i); i})(5)
5
res0: Int = 5

scala> list.filter(i => {f(i); true})(5)
0
1
2
3
4
5
res1: Int = 5
```

NthPortal replied:
I retract all complaints, and I'm happy with `fromSpecific(new View.Map(...))`

lrytz commented:
ping @joshlemer

lrytz commented:
Travis should be ignored for the time being, https://github.com/scala/scala-dev/issues/570

szeiger commented:
I see. In this case it just needs to be squashed because the first commit on its own failed.

SethTisue commented:
@joshlemer squash?

joshlemer commented:
Done! @SethTisue

SethTisue commented:
gah, sorry Josh, needs one more rebase.

I tried to use this the other day and was surprised to find it hadn't been merged yet :-)

joshlemer committed:
[11098] Add IterableOps#tapEach method

Better unit test for Vector.tapEach

tapeach returns specific iterable type

Add comments in vectortest

Fix ViewTest.tapEach

Add tapEach to iterableOnce, optimizations for strict collections

Add tests for LazyList.tapeach

Remove git merge artifacts

Fix LazyList.tapEach test

Fix LazyList#tapEach, add more and better tests

Better tests in IteratorTest.tapEach

Use 'locally' in IteratorTest#tapEach

Miscellaneous cleanup related to IterableOnce#tapEach

Remove copy of scaladoc for tapEach

Remove '(typically)' from tapEach scalado

override knownSize in IterableOnce#tapEach

joshlemer commented:
🤞 @SethTisue
